docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+let john_name = john.name
+```
+
+This is the generated JavaScript:
+
+```js
+var MySchool = require("MySchool");
+
+var john_name = MySchool.john.name;
+```
+
+External is documented in the ["Intro to `external`"](#intro-to-external)
+section. `@@bs.module` is documented
+[here](#using-functions-from-other-javascript-modules).
+
+If you want or need to use different field names on the Melange and the
+JavaScript sides, you can use the `bs.as` decorator:
+
+```ocaml
+type action = {
+  type_ : string [@bs.as "type"]
+}
+
+let action = { type_ = "ADD_USER" }
+```
+
+Which generates the JavaScript code:
+
+```js
+var action = {
+  type: "ADD_USER"
+};
+```
+
+This is useful to map to JavaScript attribute names that cannot be expressed in
+Melange, for example, where the JavaScript name we want to generate is a
+[reserved keyword](https://v2.ocaml.org/manual/lex.html#sss:keywords).
+
+It is also possible to map a Melange record to a JavaScript array by passing
+indices to the `bs.as` decorator:
+
+```ocaml
+type t = {
+  foo : int; [@bs.as "0"]
+  bar : string; [@bs.as "1"]
+}
+
+let value = { foo = 7; bar = "baz" }
+```
+
+And its JavaScript generated code:
+
+```js
+var value = [
+  7,
+  "baz"
+];
+```
+
+##### Using `Js.t` objects
+
+Alternatively to records, Melange offers another type that can be used to
+produce JavaScript objects. This type is `'a Js.t`, where `'a` is an [OCaml
+object](https://v2.ocaml.org/manual/objectexamples.html).
+
+The advantage of objects versus records is that no type declaration is needed in
+advanced, which can be helpful for prototyping or quickly generating JavaScript
+object literals.
+
+Melange provides some ways to create `Js.t` object values, as well as accessing
+the properties inside them. To create values, the `[%bs.obj]` extension is used,
+and the `##` infix operator allows to read from the object properties:
+
+```ocaml
+let john = [%bs.obj { name = "john"; age = 99 }]
+let t = john##name
+```
+
+Which generates:
+
+```js
+var john = {
+  name: "john",
+  age: 99
+};
+
+var t = john.name;
+```
+
+Note that type inference with objects is more lenient than with records. For
+example, one can create a function that operates in all the object types that
+include a field `name` that is of type string, e.g.:
+
+```ocaml
+let name_extended obj = obj##name ^ " wayne"
+
+let one = name_extended [%bs.obj { name = "john"; age = 99 }]
+let two = name_extended [%bs.obj { name = "jane"; address = "1 infinite loop" }]
+```
+
+This is possible thanks to [objects
+polymorphism](https://v2.ocaml.org/manual/objectexamples.html#s%3Apolymorphic-methods),
+and is a characteristic shared by some types in OCaml, like objects and
+polymorphic variants. The book Real World OCaml has [a good
+introduction](https://dev.realworldocaml.org/objects.html) to object
+polymorphism.
+
+##### Bind to object properties
+
+If you need to bind only to the property of a JavaScript object, you can use
+`bs.get` and `bs.set`:
+
+```ocaml
```

Comment:
we should also include an example of what this codes generates, so that people can distinguish between `foo.bar` and `foo["bar"]`